vdyp-lib/vdyp-vri/src/main/java/ca/bc/gov/nrs/vdyp/vri/VriStart.java

```diff
@@ -631,11 +649,143 @@ float quadMeanDiameterSpeciesAdjust(float x, float initialDq, float min, float m
+
+		if (baseArea <= 0) {
+			throw new StandProcessingException(
+					MessageFormat.format("Veteran layer base area ({0}) was not positive", baseArea)
+			);
+		}
+		if (treesPerHectare <= 0) {
+			throw new StandProcessingException(
+					MessageFormat.format("Veteran layer trees per hectare ({0}) was not positive", baseArea)
+			);
+		}
+		var quadMeanDiameter = BaseAreaTreeDensityDiameter.quadMeanDiameter(baseArea, treesPerHectare); // DQ(0,0)
+
+		lBuilder.adaptSpecies(veteranLayer, (sBuilder, spec) -> {
+			applyGroups(bec, spec.getGenus(), sBuilder);
+			if (primarySpecies.map(spec::equals).orElse(false)) {
+				sBuilder.addSite(iBuilder -> {
+					iBuilder.adapt(primarySite.get());
+					iBuilder.ageTotal(ageTotal);
+					iBuilder.yearsToBreastHeight(yearsToBreastHeight);
+				});
+			}
+			float specHeight = spec.getSite().flatMap(VriSite::getHeight).filter(x -> x > 0).orElse(dominantHeight);
+			float specBaseArea = spec.getFractionGenus() * baseArea;
+			float specQuadMeanDiameter = max(
+					estimateVeteranQuadMeanDiameter(spec.getGenus(), bec, specHeight), VETERAN_MIN_DQ
+			);
+			float specTreeDensity = BaseAreaTreeDensityDiameter.treesPerHectare(specBaseArea, specQuadMeanDiameter);
+			sBuilder.loreyHeight(specHeight);
+			sBuilder.baseArea(specBaseArea);
+			sBuilder.quadMeanDiameter(specQuadMeanDiameter);
+			sBuilder.treesPerHectare(specTreeDensity);
+		});
+
+		lBuilder.buildChildren();
+		var specList = lBuilder.getSpecies();
+
+		// Really crude reconciliation to the specified TPH (if any)
+		// Could improve with code similar to ROOTV01
+		if (treesPerHectare > 0) {
+			float tphSum = 0;
+			for (var spec : specList) {
+				tphSum += spec.getTreesPerHectareByUtilization().getAll();
+			}
+			float k = treesPerHectare / tphSum;
+			for (var spec : specList) {
+				float specBaseArea = spec.getBaseAreaByUtilization().getLarge();
+				float specTreesPerHectare = spec.getTreesPerHectareByUtilization().getLarge() * k;
+				float specQuadMeanDiameter = BaseAreaTreeDensityDiameter
+						.quadMeanDiameter(specBaseArea, specTreesPerHectare);
+				if (specQuadMeanDiameter < VETERAN_MIN_DQ) {
+					specQuadMeanDiameter = VETERAN_MIN_DQ;
+					specTreesPerHectare = BaseAreaTreeDensityDiameter
+							.treesPerHectare(specBaseArea, specQuadMeanDiameter);
+				}
+				spec.setTreesPerHectareByUtilization(Utils.utilizationVector(specTreesPerHectare));
+				spec.setQuadraticMeanDiameterByUtilization(Utils.utilizationVector(specQuadMeanDiameter));
+			}
+
+		}
+
+		// Sum BA and TPH
+		float tphSum = 0; // TPH(0,4)
+		float baSum = 0; // BA(0,4)
+		for (var spec : specList) {
+			tphSum += spec.getTreesPerHectareByUtilization().getAll();
+			baSum += spec.getBaseAreaByUtilization().getAll();
+		}
+
+		if (polygon.getMode().filter(mode -> mode == PolygonMode.BATC).isPresent()) {
+			// for mode 4 (with CC) input TPH was NOT available.
+			inputTreesPerHectare = tphSum;
+			treesPerHectare = tphSum;
+		}
+
+		return inputTreesPerHectare;
```

Comment:
There's this check (after the call to YUCV) in the original code - not sure where it is in your code:

      if ( abs (TPHinput- TPH(0,4))/TPHinput .gt. 0.0005)
                    LOW_DQ = .true.
      if (LOW_DQ) IER= -1